library/core/src/iter/adapters/mod.rs

```diff
@@ -187,20 +194,31 @@ where
         }
     }
 
-    fn try_fold<B, F, T>(&mut self, init: B, mut f: F) -> T
+    fn try_fold<B, F, T>(&mut self, init: B, f: F) -> T
     where
-        F: FnMut(B, Self::Item) -> T,
-        T: Try<Output = B>,
+        F: ~const FnMut(B, Self::Item) -> T + ~const Destruct,
+        T: ~const Try<Output = B>,
     {
-        self.iter
-            .try_fold(init, |acc, x| match Try::branch(x) {
+        const fn try_fold_impl<B, F, T, Item>(
+            (f, residual): &mut (F, &mut Option<Item::Residual>),
+            (acc, x): (B, Item),
+        ) -> ControlFlow<T, B>
+        where
+            F: ~const FnMut(B, Item::Output) -> T,
+            Item: ~const Try,
+            T: ~const Try<Output = B>,
+            Item::Residual: ~const Destruct,
+        {
+            match Try::branch(x) {
                 ControlFlow::Continue(x) => ControlFlow::from_try(f(acc, x)),
                 ControlFlow::Break(r) => {
-                    *self.residual = Some(r);
+                    **residual = Some(r);
```

Comment:
The double indirection is... unusual? I'm not sure if it even indicates a problem, I'm just highlighting it as something that rarely comes up.

Reply:
This has to do with the implementation of `ConstFnMutClosure`, which requires the data to be a mutable reference.